### PR TITLE
feat: complete user cli journey, add github writer

### PR DESCRIPTION
**Pull request by sailorlqh**
Example cli and output:

This example is to sync google group with a team in my own org, it's expecting `gjonathanhong` and `kevya-google` be removed.

```
> tlctl sync run -source GOOGLEGROUPS -d GITHUB -gc 'a.textproto' -uc 'b.textproto' -github-client-auth-token 'my-token'

{"time":"2024-12-12T21:39:40.748852932Z","severity":"INFO","message":"starting sync","source_group_id":"groups/04f1mdlm2alv5ca"}
{"time":"2024-12-12T21:39:40.748958629Z","severity":"INFO","message":"found the following target group IDs to sync","source_group_id":"groups/04f1mdlm2alv5ca","target_group_ids":["145597256:11735575"]}
{"time":"2024-12-12T21:39:40.748972513Z","severity":"INFO","message":"syncing target group ID","target_group_id":"145597256:11735575"}
{"time":"2024-12-12T21:39:40.748978589Z","severity":"INFO","message":"found source group ID(s) for target Group ID","target_group_id":"145597256:11735575","source_group_ids":["groups/04f1mdlm2alv5ca"]}
{"time":"2024-12-12T21:39:41.056409241Z","severity":"INFO","message":"found descendant(s) for source group ID(s)","source_group_ids":["groups/04f1mdlm2alv5ca"],"source_user_ids":["qinhang@tycho.joonix.net"]}
{"time":"2024-12-12T21:39:41.056513204Z","severity":"INFO","message":"mapped source users to target users","source_user_ids":["qinhang@tycho.joonix.net"],"target_user_ids":["sailorlqh"]}
{"time":"2024-12-12T21:39:41.056534694Z","severity":"INFO","message":"setting target group ID members to target users","target_group_id":"145597256:11735575","target_user_ids":["sailorlqh"]}
{"time":"2024-12-12T21:39:41.056555495Z","severity":"INFO","message":"fetching members for team","team_id":"145597256:11735575"}
{"time":"2024-12-12T21:39:41.48332866Z","severity":"INFO","message":"current team members","team_id":"145597256:11735575","current_member_ids":["gjonathanhong","kevya-google","sailorlqh"]}
{"time":"2024-12-12T21:39:41.483375018Z","severity":"INFO","message":"authoritative team members","team_id":"145597256:11735575","authoritative_member_ids":["sailorlqh"]}
{"time":"2024-12-12T21:39:41.483385867Z","severity":"INFO","message":"members to add","team_id":"145597256:11735575","add_member_ids":[]}
{"time":"2024-12-12T21:39:41.483394169Z","severity":"INFO","message":"members to remove","team_id":"145597256:11735575","remove_member_ids":["gjonathanhong","kevya-google"]}
```

If you look at the last line in the log, you can see the result matches our expectation.

### CODE REVIEW

File: pkg/cli/sync.go [hunk 1]
```diff
@@ -131,11 +119,16 @@ func (c *SyncCommand) Run(ctx context.Context, args []string) error {
 		return fmt.Errorf("unexpected arguments: %q", args)
 	}
 
-	if ok := slices.Contains(allowedSourceSystem, strings.ToUpper(c.source)); !ok {
+	// Convert source and destination to all caps so it matches the
+	// predefined system name const.
+	c.source = strings.ToUpper(c.source)
```
Comment (verbanicm) [inline pkg/cli/sync.go:124]: In your usage and examples above you should use the proper case then, also, the usage string can include allowed values so users dont have to guess.

e.g. https://github.com/abcxyz/guardian/blob/main/pkg/platform/config.go#L49

File: pkg/googlegroups/client.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+// NewReader creates a group reader for GoogleGroups.
+// For now we will only support read auth token from default
+// environment variable GOOGLE_APPLICATION_CREDENTIALS.
+// See the follow link for more details
+// https://cloud.google.com/docs/authentication/application-default-credentials
+//
+// You use `gcloud auth application-default login` or use
+// google-gihub-actions/auth to have the credentials written to this env var.
+func NewReader(ctx context.Context) (groupsync.GroupReader, error) {
```
Comment (verbanicm) [inline pkg/googlegroups/client.go:35]: This API feels strange to me, isnt this basically the same as NewGroupReader? Why do we need both?

Comment (sailorlqh) [reply pkg/googlegroups/client.go:35]: `NewGroupReader` in `googlegrooups.go` takes identity client and admin client, my initial thought is we will support more options for creating the identity client, and those logic will be handled in here.

So in  `cli/sync.go`, the only need to call `googlegroups.NewReader` to get the GroupReader they need.

File: pkg/cli/sync.go (hunk 1 above)
Comment (sailorlqh) [reply pkg/cli/sync.go]: Added.

File: pkg/cli/sync.go [hunk 3]
```diff
@@ -131,11 +119,16 @@ func (c *SyncCommand) Run(ctx context.Context, args []string) error {
...
+	c.destination = strings.ToUpper(c.destination)
+
+	if ok := slices.Contains(allowedSourceSystem, c.source); !ok {
```
[the remark this answers is not in the document]
Comment (sailorlqh) [reply pkg/cli/sync.go:155]: added

**Comment by sailorlqh** [comment]
Closing this PR, discussed offline for a new method. Will implement a new one.

**Pull request closed without merging**